### DISCUSSION TITLE
flax's avg_pool doesn't play well with vmap

### DISCUSSION THREAD

**Discussion by bilal2vec** [google/flax#738] [General] [answered]
Hi!

I'm using flax's avg pooling and i've run into an error when trying to use it inside a `vmap`ed function:
```
@jax.vmap
def apply(x):
  return nn.avg_pool(x, window_shape=(3, 3), strides=(2, 2), padding=((1, 1), (1, 1)))
```
```
TypeError: reduce_window got the wrong number of window_dimensions for operand: got operand shape (8, 32, 32, 3) with window_dimensions (1, 1, 3, 3, 1).
```

idk if it isn't supposed to be used in this way because the imagenet and mnist examples use it without wrapping everything in a `vmap`.

I have a colab notebook with the full stacktrace as [well](https://colab.research.google.com/drive/1_F0kALyanlHif_D5zkru2Q6WcX5jXeiS?usp=sharing)

**Comment by jheek** [comment, marked as answer]
Yes I think you are right. Pooling layers still assume there is a batch dimension. This should be an easy fix. I created an issue for this #739
